src/nspawn/nspawn.c

```diff
@@ -2594,9 +2594,24 @@ static int inner_child(
                 return -ESRCH;
         }
 
-        r = mount_systemd_cgroup_writable("", arg_unified_cgroup_hierarchy);
-        if (r < 0)
-                return r;
+	if (cg_ns_supported()) {
+		r = unshare(CLONE_NEWCGROUP);
+		if (r < 0)
+			return log_error_errno(errno, "Failed to unshare cgroup namespace");
+		r = mount_cgroups(
+				"",
+				arg_unified_cgroup_hierarchy,
+				arg_userns_mode != USER_NAMESPACE_NO,
+				arg_uid_shift,
+				arg_uid_range,
+				arg_selinux_apifs_context);
+		if (r < 0)
+			return r;
```

Comment:
No, this is not the real cause. The real cause is that `sys` is mounted read-only when `--private-veth` is used. So we are not allowed to create `/sys/fs/cgroup` which fails prior to the call you're pointing to.